TimeOfFlight/src/TOFEstimators.cc

```diff
@@ -1,633 +1,227 @@
+            }
+
+            double arcLength = getHelixArcLength( trackStates[j-1], trackStates[j] );
+            Vector3D mom = getHelixMomAtTrackState( trackStates[j-1], _bField );
+            trackLength += arcLength;
+            harmonicMom += arcLength/mom.r2();
+        }
+        harmonicMom = std::sqrt(trackLength/harmonicMom);
+
+        //////////////////////////////////////
+        // This part calculates Time of flight
+        //////////////////////////////////////
+        double timeOfFlight = 0.;
+        if( _extrapolateToEcal ){
+            if (_tofMethod == "closest"){
+                timeOfFlight = getTofClosest(cluster, track, _timeResolution/1000.);
+            }
+            else if (_tofMethod == "frankAvg"){
+                vector<CalorimeterHit*> frankHits = selectFrankEcalHits(cluster, track, _maxEcalLayer, _bField);
+                timeOfFlight = getTofFrankAvg(frankHits, track, _timeResolution/1000.);
+            }
+            else if (_tofMethod == "frankFit"){
+                vector<CalorimeterHit*> frankHits = selectFrankEcalHits(cluster, track, _maxEcalLayer, _bField);
+                timeOfFlight = getTofFrankFit(frankHits, track, _timeResolution/1000.);
+            }
+        }
+        else{
+            //define tof as average time between two SET strips
+            //if no SET hits found, tof alreasy is 0, just skip
+            TrackerHit* hitSET = getSETHit(track, _tpcOuterR);
+            if ( hitSET != nullptr ){
+                const vector<LCObject*>& simHitsSET = navigatorSET.getRelatedToObjects( hitSET );
+                if ( simHitsSET.size() >= 2 ){
+                    //It must be always 2, but just in case...
+                    if (simHitsSET.size() > 2) streamlog_out(WARNING)<<"Found more than two SET strip hits! Writing TOF as an average of the first two elements in the array."<<std::endl;
+
+                    SimTrackerHit* simHitSETFront = static_cast <SimTrackerHit*>( simHitsSET[0] );
+                    SimTrackerHit* simHitSETBack = static_cast <SimTrackerHit*>( simHitsSET[1] );
+                    double timeFront = RandGauss::shoot( simHitSETFront->getTime(), _timeResolution/1000. );
+                    double timeBack = RandGauss::shoot( simHitSETBack->getTime(), _timeResolution/1000. );
+                        timeOfFlight = (timeFront + timeBack)/2.;
+                }
+                else if (simHitsSET.size() == 1){
+                    streamlog_out(WARNING)<<"Found only one SET strip hit! Writing TOF from a single strip."<<std::endl;
+                    SimTrackerHit* simHitSET = static_cast <SimTrackerHit*>( simHitsSET[0] );
+                    timeOfFlight = RandGauss::shoot( simHitSET->getTime(), _timeResolution/1000. );
+                }
+                else{
+                    // this probably never happens...
+                    streamlog_out(WARNING)<<"Found NO simHits associated with the found SET hit! Writing TOF as 0."<<std::endl;
```

Comment:
I did not check it very thoroughly if it happens, and if so, how often...

But I checked the `DDSpacePointBuilder` code right now and in principle it can happen..
If there are >1 sim hit associated with both SET strips.

https://github.com/iLCSoft/MarlinTrkProcessors/blob/0c06ad16983b6cba545d561c5d8a5b43bcebce6d/source/Digitisers/src/DDSpacePointBuilder.cc#L327-L353

In this case, SpacePoint is there, but we weirdly can't access its time, because there are no simHits written in the relation...

_"Should this be made something more than just a log message?"_
I don't know. But they way I look at it, is that I don't want TimeEstimators single-handedly crash all production because of some SET hits... So I always constrain myself with log messages and "skip" behaviour...

I can try to study if this happens for our new Z->ff production statistics...

Do you have any further thoughts/comments on this?

Reply:
I checked on the new production of Z->2f which resulted in 8.2 mil of PFOs statistics.

The case with no SimHits happens in ~ 0.1% of pfos.
So this can happen, but this is a behaviour of the `DDSpacePointBuilder`, so we shouldn't treat it as an error?

I didn't found any cases with `1` or `>2` strips, so in principle, I can simplify if statement... But leaving it like this will probably be more robust if we change `DDSpacePointBuilder` in the future

Here is the distribution @tmadlener 
![nSEThits](https://user-images.githubusercontent.com/28848011/136232477-c64595ca-1e04-46b8-a8e9-32f8601ac069.png)

Reply:
> I don't want TimeEstimators single-handedly crash all production because of some SET hits... 

Good point. Thanks for the detailed studies. I would say a warning message is the way to go then. (maybe change the comment to say "this rarely happens" ;) )